File: BE/src/adapter/controllers/coupon.controller.js
import createError from "http-errors";
import { CouponService } from "../../core/services/index.js";
import { CheckMissingFields, FormatResult } from "../../utils/index.js";

class CouponController {
  async createCoupon(req, res, next) {
    try {
      const { code, discountAmount, usageLimit } = req.body;
      const isActive = String(req.body.isActive).toLowerCase() === "true";

      CheckMissingFields({
        code,
        discountAmount,
        usageLimit,
      });

      const newCoupon = {
        code,
        discountAmount,
        usageLimit,
        isActive,
      };

      const data = await CouponService.CreateCoupon(newCoupon);
      res.status(201).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }

  async getCouponDetail(req, res, next) {
    const code = req.params.id;

    try {
      const data = await CouponService.GetCouponByCode(code);
      res.status(200).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }

  async getListCoupon(req, res, next) {
    try {
      const data = await CouponService.GetAllCoupons();
      res.status(200).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }

  async filterCoupon(req, res, next) {
    const { sortBy, filter } = req.body || {};

    try {
      const data = await CouponService.GetCouponByFilter(filter, sortBy);
      res.status(200).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }

  async updateCoupon(req, res, next) {
    const code = req.params.code;

    try {
      const updatedCoupon = req.body;
      const data = await CouponService.UpdateCoupon(code, updatedCoupon);
      res.status(200).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }

  async deleteCoupon(req, res, next) {
    const code = req.params.code;

    try {
      const data = await CouponService.DeleteCoupon(code);
      res.status(200).json(FormatResult("success", data));
    } catch (err) {
      next(createError(400, err));
    }
  }
}

export default new CouponController();

File: FE/ecommerce/lib/features/auth/controllers/coupon_controller.dart
import 'package:ecommerce/utils/http/http_client.dart';
class CouponController {
  Future<Map<String, dynamic>> getCoupons() async {
    final response = await CHttpHelper.get('coupons');
    return response;
  }

  // Future<Map<String, dynamic>> filterCoupons({
  //   Map<String, dynamic>? filter,
  //   Map<String, dynamic>? sortBy,
  // }) async {
  //   final body = <String, dynamic>{};

  //   if (filter != null) body['filter'] = filter;
  //   if (sortBy != null) body['sortBy'] = sortBy;

  //   print("=========body filter: $body" );

  //   final response = await CHttpHelper.post('coupons/filter', body);
  //   return response;
  // }

  // Future<Map<String, dynamic>> getCouponDetail(String id) async {
  //   final response = await CHttpHelper.get('coupons/$id');
  //   return response;
  // }

  Future<Map<String, dynamic>> createCoupon(Map<String, dynamic> couponData) async {
    final response = await CHttpHelper.post('coupons', couponData, withAuth: true);
    return response;
  }

  Future<Map<String, dynamic>> updateCoupon(String id, Map<String, dynamic> couponData) async {
    final response = await CHttpHelper.put('coupons/$id', couponData, withAuth: true);
    return response;
  }

  Future<Map<String, dynamic>> deleteCoupon(String code) async {
    final response = await CHttpHelper.delete('coupons/$code', withAuth: true);
    return response;
  }
}